src/Core/Helpers.cs

```diff
@@ -28,39 +30,44 @@ internal static IEnumerable<Assembly> GetReferencingAssemblies()
             return assemblies;
         }
 
-        internal static async Task HandleProxy(HttpContext context, string uri, Func<HttpContext, Exception, Task> onFailure = null)
+        internal static async Task ExecuteProxyOperation(HttpContext context, string uri, ProxyOptions options = null)
         {
             try
             {
-                var proxiedResponse = await context.SendProxyHttpRequest(uri).ConfigureAwait(false);
-                await context.CopyProxyHttpResponse(proxiedResponse).ConfigureAwait(false);
+                var proxiedRequest = context.CreateProxiedHttpRequest(uri, options?.ShouldAddForwardedHeaders ?? true);
+
+                options?.BeforeSend?.Invoke(context, proxiedRequest);
+                var proxiedResponse = await context.SendProxiedHttpRequest(proxiedRequest).ConfigureAwait(false);
+
+                options?.AfterReceive?.Invoke(context, proxiedResponse);
+                await context.WriteProxiedHttpResponse(proxiedResponse).ConfigureAwait(false);
             }
             catch (Exception e)
             {
-                if (onFailure == null)
+                if (options?.HandleFailure == null)
                 {
                     // If the failures are not caught, then write a generic response.
-                    context.Response.StatusCode = 500;
+                    context.Response.StatusCode = 502;
                     await context.Response.WriteAsync($"Request could not be proxied.\n\n{e.Message}\n\n{e.StackTrace}.").ConfigureAwait(false);
                     return;
                 }
 
-                await onFailure(context, e).ConfigureAwait(false);
+                options?.HandleFailure?.Invoke(context, e);

```

Comment:
not needed Elvis operator here, we already check, that `options`,  and `HandleFailure` not null

Reply:
Ahh, good catch...will fix.